File: src/main/java/org/railwaystations/rsapi/adapter/mail/SmtpMailer.java
package org.railwaystations.rsapi.adapter.mail;

import org.apache.commons.lang3.StringUtils;
import org.railwaystations.rsapi.domain.port.out.Mailer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import javax.mail.*;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeBodyPart;
import javax.mail.internet.MimeMessage;
import javax.mail.internet.MimeMultipart;
import java.util.Properties;

@Service
public class SmtpMailer implements Mailer {

    private static final Logger LOG = LoggerFactory.getLogger(SmtpMailer.class);
    private final MailerConfig config;

    private Session session;

    public SmtpMailer(final MailerConfig config) {
        this.config = config;
        if (StringUtils.isNoneBlank(config.getHost())) {
            final Properties properties = System.getProperties();
            properties.setProperty("mail.smtp.host", config.getHost());
            properties.setProperty("mail.smtp.auth", "true");
            properties.setProperty("mail.smtp.port", config.getPort());
            properties.setProperty("mail.smtp.ssl.enable", "true");
            properties.setProperty("mail.smtp.timeout", "10000");
            properties.setProperty("mail.smtp.connectiontimeout", "10000");

            session = Session.getInstance(properties, new UsernamePasswordAuthenticator(config.getUser(), config.getPasswd()));
        }
    }

    @Override
    public void send(final String to, final String subject, final String text) {
        if (session == null) {
            LOG.info("Mailer not initialized, can't send mail to {} with subject {} and body {}", to, subject, text);
            return;
        }
        try {
            LOG.info("Sending mail to {}", to);
            final MimeMessage message = new MimeMessage(session);
            message.setFrom(new InternetAddress(config.getFrom()));
            message.addRecipient(Message.RecipientType.TO, new InternetAddress(to));
            message.setSubject(subject);

            final Multipart multipart = new MimeMultipart();

            final MimeBodyPart textBodyPart = new MimeBodyPart();
            textBodyPart.setText(text);
            multipart.addBodyPart(textBodyPart);

            message.setContent(multipart);
            Transport.send(message);
            LOG.info("Mail sent");
        } catch (final Exception e) {
            throw new RuntimeException("Unable to send mail", e);
        }
    }

    private static class UsernamePasswordAuthenticator extends Authenticator {
        private final String user;
        private final String passwd;

        private UsernamePasswordAuthenticator(final String user, final String passwd) {
            this.user = user;
            this.passwd = passwd;
        }

        @Override
        protected PasswordAuthentication getPasswordAuthentication() {
            return new PasswordAuthentication(user, passwd);
        }
    }
}

File: service/src/main/kotlin/org/railwaystations/rsapi/adapter/mail/SmtpMailer.kt
package org.railwaystations.rsapi.adapter.mail

import jakarta.mail.Authenticator
import jakarta.mail.Message
import jakarta.mail.PasswordAuthentication
import jakarta.mail.Session
import jakarta.mail.Transport
import jakarta.mail.internet.InternetAddress
import jakarta.mail.internet.MimeBodyPart
import jakarta.mail.internet.MimeMessage
import jakarta.mail.internet.MimeMultipart
import org.apache.commons.lang3.StringUtils
import org.railwaystations.rsapi.core.ports.Mailer
import org.railwaystations.rsapi.utils.Logger
import org.springframework.stereotype.Service

@Service
class SmtpMailer(private val config: MailerConfig) : Mailer {
    private lateinit var session: Session
    private val log by Logger()

    init {
        if (StringUtils.isNoneBlank(config.host)) {
            val properties = System.getProperties().apply {
                setProperty("mail.smtp.host", config.host)
                setProperty("mail.smtp.auth", "true")
                setProperty("mail.smtp.port", config.port)
                setProperty("mail.smtp.ssl.enable", config.ssl.toString())
                setProperty("mail.smtp.timeout", "10000")
                setProperty("mail.smtp.connectiontimeout", "10000")
            }

            session = Session.getInstance(properties, UsernamePasswordAuthenticator(config.user, config.passwd))
        }
    }

    override fun send(to: String, subject: String, text: String) {
        try {
            log.info("Sending mail to {}", to)

            val multipart = MimeMultipart().apply {
                val textBodyPart = MimeBodyPart()
                textBodyPart.setText(text)
                addBodyPart(textBodyPart)
            }

            val message = MimeMessage(session).apply {
                setFrom(InternetAddress(config.from))
                addRecipient(Message.RecipientType.TO, InternetAddress(to))
                setSubject(subject)
                setContent(multipart)
            }

            Transport.send(message)
            log.info("Mail sent")
        } catch (e: Exception) {
            throw RuntimeException("Unable to send mail", e)
        }
    }

    private class UsernamePasswordAuthenticator(private val user: String, private val passwd: String) :
        Authenticator() {
        override fun getPasswordAuthentication(): PasswordAuthentication {
            return PasswordAuthentication(user, passwd)
        }
    }
}
